fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/pdagg/PushDownAggregateCollector.java

```diff
@@ -0,0 +1,521 @@
+        if (isInvalid(optExpression, context)) {
+            return visit(optExpression, context);
+        }
+
+        List<PushDownAggregateCollector> collectors = Lists.newArrayList();
+        LogicalUnionOperator union = (LogicalUnionOperator) optExpression.getOp();
+        for (int i = 0; i < optExpression.getInputs().size(); i++) {
+            List<ColumnRefOperator> childOutput = union.getChildOutputColumns().get(i);
+            Map<ColumnRefOperator, ScalarOperator> rewriteMap = Maps.newHashMap();
+            Preconditions.checkState(childOutput.size() == union.getOutputColumnRefOp().size());
+            for (int k = 0; k < union.getOutputColumnRefOp().size(); k++) {
+                rewriteMap.put(union.getOutputColumnRefOp().get(k), childOutput.get(k));
+            }
+
+            ReplaceColumnRefRewriter rewriter = new ReplaceColumnRefRewriter(rewriteMap);
+            AggregateContext childContext = new AggregateContext();
+            childContext.origAggregator = context.origAggregator;
+            childContext.aggregations.putAll(context.aggregations);
+            childContext.aggregations.replaceAll((k, v) -> (CallOperator) rewriter.rewrite(v));
+
+            childContext.groupBys.putAll(context.groupBys);
+            childContext.groupBys.replaceAll((k, v) -> rewriter.rewrite(v));
+            childContext.pushPaths.addAll(context.pushPaths);
+            childContext.pushPaths.add(i);
+
+            PushDownAggregateCollector collector = new PushDownAggregateCollector(this.taskContext);
+            collectors.add(collector);
+            collector.process(optExpression.inputAt(i), childContext);
+        }
+
+        // collect push down aggregate context
+        List<List<AggregateContext>> allChildRewriteContext = Lists.newArrayList();
+        for (PushDownAggregateCollector childCollector : collectors) {
+            if (childCollector.allRewriteContext.containsKey(context.origAggregator)) {
+                allChildRewriteContext.add(childCollector.allRewriteContext.get(context.origAggregator));
+                childCollector.allRewriteContext.remove(context.origAggregator);
+            }
+        }
+
+        // merge other rewrite context
+        for (PushDownAggregateCollector collector : collectors) {
+            Preconditions.checkState(
+                    !CollectionUtils.containsAny(allRewriteContext.keySet(), collector.allRewriteContext.keySet()));
+            allRewriteContext.putAll(collector.allRewriteContext);
+        }
+
+        // none aggregate can push down to union children
+        if (allChildRewriteContext.isEmpty() || allChildRewriteContext.size() != collectors.size()) {
+            return null;
+        }
+
+        Set<ColumnRefOperator> checkGroupBys = new HashSet<>(context.groupBys.keySet());
+        Set<ColumnRefOperator> checkAggregations = new HashSet<>(context.aggregations.keySet());
+
+        for (List<AggregateContext> childContexts : allChildRewriteContext) {
```

Comment:
Don't understand this logic.  does this allChildRewriteContext have the order?

Reply:
None order, for check that the aggregation can be pushed down to all union children, we must promise push down to all or don't push down to anyone